File: Source/ExcelDna.CustomRegistration.VBExample/ExampleAddIn.vb
Imports ExcelDna.Integration

Public Class ExampleAddIn
    Implements IExcelAddIn

    Public Sub AutoOpen() Implements IExcelAddIn.AutoOpen
        ExcelIntegration.RegisterUnhandledExceptionHandler(Function(ex) "!!! ERROR: " + ex.ToString())

        Dim conversionConfig = New ParameterConversionConfiguration()
        conversionConfig.AddParameterConversion(AddressOf ParameterConversions.OptionalConversion)
        
        Registration.GetExcelFunctions() _
                    .ProcessParameterConversions(conversionConfig) _
                    .ProcessParamsRegistrations() _
                    .RegisterFunctions()

        ' Could add Async too...
    End Sub
    
    Public Sub AutoClose() Implements IExcelAddIn.AutoClose

    End Sub

End Class

File: Source/Samples/Registration.Sample.FSharp/ExampleAddIn.fs
namespace Registration.Samples.FSharp

open System
open ExcelDna.Integration
open ExcelDna.Registration
open ExcelDna.Registration.FSharp

type FsAsyncAddIn () =
    interface IExcelAddIn with
        member this.AutoOpen ()  = 
            // The overload selection and delegate conversions performed by F# are not intuitive.
            let paramConvertConfig = ParameterConversionConfiguration()
                                        .AddParameterConversion( 
                                            (fun (typ : Type) (reg : ExcelParameterRegistration) -> FsParameterConversions.FsOptionalParameterConversion typ reg),
                                             null)

            ExcelRegistration.GetExcelFunctions ()
            |> fun fns -> ParameterConversionRegistration.ProcessParameterConversions (fns, paramConvertConfig)
            |> FsAsyncRegistration.ProcessFsAsyncRegistrations
            |> AsyncRegistration.ProcessAsyncRegistrations
            |> MapArrayFunctionRegistration.ProcessMapArrayFunctions
            |> ExcelRegistration.RegisterFunctions
        
        member this.AutoClose () = ()
